### PR TITLE
Movable HTTPClient and fixing WiFiClient copy

### LINKED ISSUE #5734: Crash when Debug port is Disabled - HTTPClient destructor issue?
### Basic Infos

- [x] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [x] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [ ] I have tested that the issue is present in current master branch (aka latest git).
- [x] I have searched the issue tracker for a similar issue.
- [ ] If there is a stack dump, I have decoded it.
- [x] I have filled out all fields below.

#### Platform

- Hardware:				[ESP-12]
- Core Version:			[just released 2.5.0]
- Development Env:		[Arduino IDE 1.8.8]
- Operating System:		[Ubuntu]

### Settings in IDE

- Module:  			[Generic ESP8266 Module]
- Flash Mode:			[dout]
- Flash Size: 		[4MB]
- lwip Variant:   [v2 Lower Memory]
- Reset Method:		[ck]
- Flash Frequency:	[40Mhz]
- CPU Frequency:		[80Mhz]
- Upload Using:		[SERIAL]
- Upload Speed:   [115200]

### Problem Description

In Arduino IDE when `Debug port` is set to `Disabled` the ESP8266 reboots (most probably due to a crash) a short while after fetching JSON data via https, parsing it and displaying it on a LED matrix. However, if I set the `Debug port` to `Serial` then everything works properly. My application does not use any debugging stuff. It does not even use the Serial stuff, does not call `Serial.begin()`, nothing of that.

The `Debug level` is set to `None` all the time so there's no reason why `Debug port` setting should be significant or should affect the application in any way, IMHO.

I cannot debug this because when I enable debugging and thus set the Debug port to Serial the issue disappears :-)

EDIT: here's the minimal code:
```
#include <ESP8266WiFi.h>
#include <ESP8266HTTPClient.h>
#include <WiFiClientSecureBearSSL.h>

const int led = 5;

void setup(void)
{
  pinMode(led, OUTPUT);
  for(byte i=0; i<5; i++) {
    digitalWrite(led, HIGH);
    delay(150);
    digitalWrite(led, LOW);
    delay(200);
  }
  
  WiFi.mode(WIFI_STA);
  // Wait for connection
  while (WiFi.status() != WL_CONNECTED) {
    delay(250);
  }
}

void displayData()
{
  digitalWrite(led, HIGH);
  HTTPClient http;
  std::unique_ptr<BearSSL::WiFiClientSecure>client(new BearSSL::WiFiClientSecure);
  client->setInsecure();
  if (http.begin(*client, "https://jigsaw.w3.org/HTTP/connection.html")) {
    if (http.GET() == 200) {
      String payload = http.getString();
    }
    http.end();
  }
  digitalWrite(led, LOW);
}

void loop(void)
{
  static unsigned long mil = 0;
  if (millis() - mil > 5000) {   
    displayData();
    mil = millis();
  }
}
```

It's easy to reproduce that mere `Debug port` setting turns application crashing to not crashing and vice-versa.

It's obvious from the source code above that the LED should blink 5 times at start, then wait for up to 5 seconds, and then blink (when the HTTPS request is invoked) once every five seconds.
What you observe is that immediately after the longer HTTPS blink there's the boot-up fast 5 blinks. That indicates the crash&reboot.

EDIT2: you might notice that my MCVE is an almost verbatim copy of the "_BasicHttpsClient_" example that is provided with Arduino core. Interesting is that the stock BasicHttpsClient does not crash. What is different? Well, the only interesting difference is the order of HTTPClient and BearSL client instantiation. And that's actually it! If you switch the order of those two lines in the BasicHttpsClient example (so that `HTTPClient` is first, the `std::unique_ptr` is second) it will start crashing as well, this time with a stack trace:

```
[SETUP] WAIT 4...
[SETUP] WAIT 3...
[SETUP] WAIT 2...
[SETUP] WAIT 1...
[HTTPS] begin...
[HTTPS] GET...

Exception (9):
epc1=0x40206cb9 epc2=0x00000000 epc3=0x00000000 excvaddr=0x000001cb depc=0x00000000

>>>stack>>>

ctx: cont
sp: 3ffffd60 end: 3fffffc0 offset: 01a0
3fffff00:  00000001 3ffeecd8 3ffefd14 40202842  
3fffff10:  00000000 00000000 3ffefd14 3ffefa94  
3fffff20:  0000000f 0000000d 000001bb 40001388  
3fffff30:  3ffefe1c 0000001f 00000015 3ffefac4  
3fffff40:  0000000f 00000005 3ffefadc 0000000f  
3fffff50:  00000000 3ffefb14 0000001f 00000011  
3fffff60:  3ffefb3c 0000000f 00000000 00000000  
3fffff70:  00000000 00000000 ffffffff 3fffef00  
3fffff80:  00000000 00000000 00000000 00000000  
3fffff90:  00000000 00000000 3ffeecd8 40202761  
3fffffa0:  3fffdad0 00000000 3ffeed20 402082b4  
3fffffb0:  feefeffe feefeffe 3ffe8508 401009e9  
<<<stack<<<

 ets Jan  8 2013,rst cause:2, boot mode:(1,7)


 ets Jan  8 2013,rst cause:4, boot mode:(1,7)

wdt reset
```

### LINKED ISSUE #8231: Unable to set HTTPClient to a std::optional
### Basic Infos

- [X] This issue complies with the [issue POLICY doc](https://github.com/esp8266/Arduino/blob/master/POLICY.md).
- [X] I have read the documentation at [readthedocs](https://arduino-esp8266.readthedocs.io/en/latest) and the issue is not addressed there.
- [ ] I have tested that the issue is present in current master branch (aka latest git).
- [X] I have searched the issue tracker for a similar issue.
- [X] If there is a stack dump, I have decoded it.
- [X] I have filled out all fields below.

#### Platform

- Hardware:					ESP-12
- Core Version:			3.0.1
- Development Env:		Platformio
- Operating System:		Ubuntu

### Problem Description

We are unable to set a HTTPClient to a std::optional<HTTPClient> that already exists. I have no idea why, the compiler generates hundreds of lines of error all related to the fact that `std::optional<HTTPClient>(std::optional<HTTPClient>&&)` is implicitly deleted. So right now it's impossible to set a HTTPClient to a std::optional, and I have no idea how to make it work, or provide a PR for this (although I'm willing if pointed in the right direction). But this seems a bug. And is something we actually need (we memset the unused 4kb sys stack to zero and cast it to a struct that contains a bunch of classes, and the easier way to properly initialize them is to wrap them in an optional).

### [MCVE](https://stackoverflow.com/help/mcve) Sketch

```cpp

#include <ESP8266HTTPClient.h>
#include <optional>

void setup() {
    std::optional<HTTPClient> http;
    http = std::make_optional(HTTPClient());
}

void loop() {

}

```

### Debug Messages

It's a huge compiler error: https://pastebin.com/2NZzvrgY

The core of it seems this:
```
include/loop.hpp:189:57: error: use of deleted function 'std::optional<HTTPClient>& std::optional<HTTPClient>::operator=(std::optional<HTTPClient>&&)'
  189 |       this->data->http = std::make_optional(HTTPClient());

.platformio/packages/toolchain-xtensa@2.100200.0/xtensa-lx106-elf/include/c++/10.2.0/optional:659:11: note: 'std::optional<HTTPClient>& std::optional<HTTPClient>::operator=(std::optional<HTTPClient>&&)' is implicitly deleted because the default definition would be ill-formed:
  659 |     class optional
      |           ^~~~~~~~

.platformio/packages/toolchain-xtensa@2.100200.0/xtensa-lx106-elf/include/c++/10.2.0/optional:659:11: error: use of deleted function 'std::_Enable_copy_move<false, false, false, false, _Tag>& std::_Enable_copy_move<false, false, false, false, _Tag>::operator=(std::_Enable_copy_move<false, false, false, false, _Tag>&&) [with _Tag = std::optional<HTTPClient>]'

.platformio/packages/toolchain-xtensa@2.100200.0/xtensa-lx106-elf/include/c++/10.2.0/bits/enable_special_members.h:306:5: note: declared here
  306 |     operator=(_Enable_copy_move&&) noexcept                         = delete;

.platformio/packages/toolchain-xtensa@2.100200.0/xtensa-lx106-elf/include/c++/10.2.0/optional: In instantiation of 'constexpr std::optional<typename std::decay<_Tp>::type> std::make_optional(_Tp&&) [with _Tp = HTTPClient; typename std::decay<_Tp>::type = std::decay<HTTPClient>::type]':
include/loop.hpp:189:57:   required from here .platformio/packages/toolchain-xtensa@2.100200.0/xtensa-lx106-elf/include/c++/10.2.0/optional:1207:62: error: no matching function for call to 'std::optional<HTTPClient>::optional(<brace-enclosed initializer list>)'
 1207 |     { return optional<decay_t<_Tp>> { std::forward<_Tp>(__t) }; }

```

And it goes on and on.

### PR DESCRIPTION
**Pull request by mcspr**
- =default for default ctor, destructor, move ctor and the assignment move
- use std::unique_ptr<WiFiClient> instead of raw pointer
- implement clone() for the WiFiClient so it's safe to copy the WiFiClientSecure, and not accidentally call the WiFiClient copy instead
- replace headers pointer array with unique_ptr<T[]> to simplify the move operations
- substitute userAgent with the default one when it is empty

Allow HTTPClient to be placed inside of movable classes (e.g. std::optional) or to be returned from functions. Class logic stays as-is, only the underlying member types are slightly changed.

replaces #8236
resolves #8231
@paulocsanz 

*edit:*
also should resolve #5734 with the WiFiClient changes

### CODE REVIEW

**Comment by paulocsanz** [comment]
This seems the perfect solution!

**Comment by paulocsanz** [comment]
There is a tiny breaking change, that setting userAgent to String() doesn't work anymore, the only way to ensure seems std::optional<String>. But it is overkill and I don't even know if a empty userAgent is supported in HTTP spec.

**Comment by mcspr** [comment]
> There is a tiny breaking change, that setting userAgent to an empty string doesn't work anymore, the only way to ensure seems std::optional. But it is overkill and I don't even know if a empty userAgent is supported in HTTP spec.

Technically... it is, but depends on the client / server reading those. Replaced with a static String that is default value for userAgent.

**Comment by paulocsanz** [comment]
Yeah it's not a big deal, I don't feel confident enough to give a position on the tradeoff empty user agent after move vs changing empty user agents for the default. Both seem to have issues when taking into account backwards compatibility. But technically move operations allow that in cpp, it's a valid but indeterminate value. I don't like taking decisions only on the "tecnically" side of things. But it's good to have that on the table before the "last word" on the subject.

Again, I'm not really a person with reputation in the community and knowledge on the subject to give a strong position, it was just my 2 cents.

File: libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h [hunk 1]
```diff
@@ -26,11 +26,12 @@
 #ifndef ESP8266HTTPClient_H_
 #define ESP8266HTTPClient_H_
 
-#include <memory>
 #include <Arduino.h>
 #include <StreamString.h>
 #include <WiFiClient.h>
 
+#include <memory>
```
Comment (paulocsanz) [inline libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h:33]: Why is this include needed? There already is a std::unique_ptr in the class (`_payload`), so it seems it's already included.

**Comment by mcspr** [comment]
I did state that the code did not change the (observable) behaviour, and it did. Note the line with String::reserve above that uses userAgent length. Plus, setting userAgent to something with length, then resetting back to empty user agent set it to default string and not the empty one as before.

"technically" referred to the [http rfc(s) and some discussions](https://github.com/whatwg/fetch/issues/332) I found related to the question you raised - could user-agent be empty, or is it allowed for the header field-value to be empty in browsers, http servers, standalone http clients, etc.

File: libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h (hunk 1 above)
Comment (mcspr) [reply libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h:33]: see L29 just above, it was before Arduino.h previously

Comment (paulocsanz) [reply libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h:33]: Oh sorry, didn't see that!

**Comment by paulocsanz** [comment]
In the worst case, wrapping `_userAgent` with a class that sets to default when moved out would solve it. If we are trying to avoid depending on manually writing the `HTTPClient` move operations to make extensibility easier. But maybe we are creating so many independent changes to have something that ends up being equally complex in the long run.

But the boilerplate here could be generalized for all libraries in the ecosystem to use. This `NonOwningPtr<T>` certainly is frequently needed, and a `DefaultOnMove<T, const T value>` (or getting a function pointer to generate `T` on construction) could be used here too and generalized. My dream would also be a internal class that is basically a `std::variant<String, __FlashStringHelper*>` with `String` API so we can avoid needless "persistent" allocations where we know the value at compile time. And the usage of `StringView` where `String` API is needed instead of forcing an allocation. But I digress.

**Comment by mcspr** [comment]
Once again with the user-agent, this may be slightly safer approach
```diff
diff --git a/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h b/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h
index 3f6ea010..b8885176 100644
--- a/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h
+++ b/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.h
@@ -284,6 +284,35 @@ protected:
         WiFiClient* _ptr = nullptr;
     };

+    struct UserAgent {
+        UserAgent() = delete;
+        explicit UserAgent(const char* ptr, size_t length) :
+            _ptr(ptr),
+            _length(length)
+        {}
+
+        template <typename T>
+        UserAgent& operator=(T&& value) {
+            _string = std::forward<T>(value);
+            _ptr = _string.c_str();
+            _length = _string.length();
+            return *this;
+        }
+
+        size_t length() const {
+            return _length;
+        }
+
+        const char* c_str() const {
+            return _ptr;
+        }
+
+    private:
+        String _string;
+        const char* _ptr;
+        size_t _length;
+    };
+
     bool beginInternal(const String& url, const char* expectedProtocol);
     void disconnect(bool preserveClient = false);
     void clear();
@@ -307,8 +336,9 @@ protected:
     String _headers;
     String _base64Authorization;

-    static const String defaultUserAgent;
-    String _userAgent = defaultUserAgent;
+    static const char defaultUserAgent[];
+    static const size_t defaultUserAgentLength;
+    UserAgent _userAgent { defaultUserAgent, defaultUserAgentLength };

     /// Response handling
     std::unique_ptr<RequestArgument[]> _currentHeaders;
```
And setup like this
```diff
diff --git a/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp b/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp
index 5aca23a8..8a3a6252 100644
--- a/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp
+++ b/libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp
@@ -35,8 +35,8 @@ static_assert(!std::is_copy_constructible_v<HTTPClient>, "");
 static_assert(std::is_move_constructible_v<HTTPClient>, "");
 static_assert(std::is_move_assignable_v<HTTPClient>, "");

-static const char defaultUserAgentPstr[] PROGMEM = "ESP8266HTTPClient";
-const String HTTPClient::defaultUserAgent = defaultUserAgentPstr;
+const char HTTPClient::defaultUserAgent[] PROGMEM = "ESP8266HTTPClient";
+const size_t HTTPClient::defaultUserAgentLength = sizeof(HTTPClient::defaultUserAgentLength) - 1;

 static int StreamReportToHttpClientReport (Stream::Report streamSendError)
 {
@@ -901,7 +901,7 @@ bool HTTPClient::sendHeader(const char * type)
         header += String(_port);
     }
     header += F("\r\nUser-Agent: ");
-    header += _userAgent;
+    header.concat(_userAgent.c_str(), _userAgent.length());

     if (!_useHTTP10) {
         header += F("\r\nAccept-Encoding: identity;q=1,chunked;q=0.1,*;q=0");
```

Since this does not involve global defaultUserAgent string and directly concats from the pointer with a known length. Notably, this also raises the question why flashstrings were not something like `struct fstr { const char* ptr; size_t length; }` in the first place so we don't have to strlen_P needlessly.

**Review by d-a-v — APPROVED** [review]
Looks good, Nothing jumps at me.
I see this change as an encapsulating class for WiFiClient which forbids copy constructor, allows and implements fake move constructor (of a pointer) and calling ->stop() on destruction.
Thanks for the time spent on this !

**Comment by mcspr** [comment]
> I see this change as an encapsulating class for WiFiClient which forbids copy constructor, allows and implements fake move constructor (of a pointer) and calling ->stop() on destruction.

Pretty much, yes. Looking at this again though... consider that WiFiClient implements WiFiClient(WiFiClient&&) and operator=(WiFiClient&&), in a way that the underlying connection is simply transfered from one object to another:
- httpclient's `_client` becomes `unique_ptr<WiFiClient>`
- instead of `_client = client` in the `begin(WiFiClient& client)`, it becomes  `_client = std::make_unique<WiFiClient>(client);`, also solving the issue with the order of object declaration that is mentioned in the TODO
- NonOwningClientPtr is removed as unneeded piece of code

**Comment by mcspr** [comment]
> instead of _client = client in the begin(WiFiClient& client), it becomes _client = std::make_unique<WiFiClient>(client);

Scratch that. My original thought was that WiFiClient already handles the virtual copies... it does not, and it 'slices' when secure connection object is copied. Meaning, it should have something like `virtual std::unique_ptr<WiFiClient> T::clone() const` for each `T` implementing WiFiClient stuff - currently that is WiFiClientSecureCtx and the WiFiClientSecure itself (that wraps the ctx). 

imo that would make more sense in the current implementation, and possible changes further to simplify the wificlient<->wificlientsecure<->clientcontext relations. Will clean up the wip and push here
(...to also possibly fix the https://github.com/esp8266/Arduino/issues/5734)

**Review by d-a-v — APPROVED** [review]
This is not the same PR since last approval.
https://github.com/esp8266/Arduino/pull/8237/files#diff-15722e0b5cf418bd34281d1ee61ad5b3e59c1ed22d9c33317ea98bc84b1a20abR104 is now the core of the changes.

Maybe more comments on what's left to do around these new `clone` methods would be useful to help refactor allocation for both ssl and non-ssl context api.

(*edit* tested to work with `BasicHttpsClient` example)

**Comment by paulocsanz** [comment]
It seemed everybody was satisfied with the old changes, are the new ones that need review and have a few TODOs the best way? I'm asking as someone with interest in merging this as fast as we can, but I understand the need for being thorough before stabilizing things.

**Comment by mcspr** [comment]
@paulocsanz The delay is not intentional, but idk what answer you expect for the comment above :/ There is a solution for the issue you reported and confirmed this worked for your use-case, and this is definitely will be a part of the next release, so no need to worry about that. Also note that I am merging this request, so the whole blame is on be atm.

> And it's not a trivial change, it has brought other questions to the table.

I'd ask for specifics here, not the general feeling.

**Pull request merged**